registry/storage/garbagecollect.go

```diff
@@ -60,26 +60,37 @@ func MarkAndSweep(ctx context.Context, storageDriver driver.StorageDriver, regis
 			return fmt.Errorf("unable to convert ManifestService into ManifestEnumerator")
 		}
 
+		unmarkedManifests := make(map[digest.Digest]struct{})
+		var manifestList []digest.Digest
+
 		err = manifestEnumerator.Enumerate(ctx, func(dgst digest.Digest) error {
 			if opts.RemoveUntagged {
-				// fetch all tags where this manifest is the latest one
 				tags, err := repository.Tags(ctx).Lookup(ctx, distribution.Descriptor{Digest: dgst})
 				if err != nil {
 					return fmt.Errorf("failed to retrieve tags for digest %v: %v", dgst, err)
 				}
 				if len(tags) == 0 {
-					emit("manifest eligible for deletion: %s", dgst)
-					// fetch all tags from repository
-					// all of these tags could contain manifest in history
-					// which means that we need check (and delete) those references when deleting manifest
-					allTags, err := repository.Tags(ctx).All(ctx)
-					if err != nil {
-						return fmt.Errorf("failed to retrieve tags %v", err)
-					}
-					manifestArr = append(manifestArr, ManifestDel{Name: repoName, Digest: dgst, Tags: allTags})
-					return nil
+					unmarkedManifests[dgst] = struct{}{}
+				} else {
+					manifestList = append(manifestList, dgst)
 				}
+			} else {
+				manifestList = append(manifestList, dgst)
 			}
+			return nil
+		})
+		// In certain situations such as unfinished uploads, deleting all
+		// tags in S3 or removing the _manifests folder manually, this
+		// error may be of type PathNotFound.
+		//
+		// In these cases we can continue marking other manifests safely.
+		if _, ok := err.(driver.PathNotFoundError); ok {
+			return nil
+		}
+
+		for i := 0; i < len(manifestList); i++ {
```

Comment:
Loop over all reachable manifests. Note that we are potentially adding newly discovered reachable manifests to `manifestList` as we go. If `opts.RemoveUntagged` we are in effect doing a breadth first search from our starting set of all tagged manifests.